src/runtime_src/xdp/profile/plugin/aie_trace/aie_trace_plugin.cpp

```diff
@@ -721,36 +721,72 @@ namespace xdp {
-    bool     isPLIO = ((db->getStaticInfo()).getNumTracePLIO(deviceId)) ? true : false;
+    // Ensure trace buffer size is appropriate
+    uint64_t aieTraceBufSize = GetTS2MMBufSize(true /*isAIETrace*/);
+    bool isPLIO = (db->getStaticInfo()).getNumTracePLIO(deviceId) ? true : false;
 
-#if 0
-    uint8_t memIndex = 0;
-    if(isPLIO) { // check only one memory for PLIO and for GMIO , assume bank 0 for now
-      memIndex = deviceIntf->getAIETs2mmMemIndex(0);
-    }
+    // First, check against memory bank size
+    // NOTE: Check first buffer for PLIO; assume bank 0 for GMIO
+    uint8_t memIndex = isPLIO ? deviceIntf->getAIETs2mmMemIndex(0) : 0;
     Memory* memory = (db->getStaticInfo()).getMemory(deviceId, memIndex);
-    if(nullptr == memory) {
-      std::string msg = "Information about memory index " + std::to_string(memIndex)
-                         + " not found in given xclbin. So, cannot check availability of memory resource for device trace offload.";
-      xrt_core::message::send(xrt_core::message::severity_level::warning, "XRT", msg);
-      return;
+    if (memory != nullptr) {
+      uint64_t fullBankSize = memory->size * 1024;
+      
+      if ((fullBankSize > 0) && (aieTraceBufSize > fullBankSize)) {
+        aieTraceBufSize = fullBankSize;
+        std::string msg = "Requested AIE trace buffer is too big for memory resource. Limiting to " + std::to_string(fullBankSize) + "." ;
+        xrt_core::message::send(xrt_core::message::severity_level::warning, "XRT", msg);
+      }
     }
-    uint64_t aieMemorySz = (((db->getStaticInfo()).getMemory(deviceId, memIndex))->size) * 1024;
-    if(aieMemorySz > 0 && aieTraceBufSz > aieMemorySz) {
-      aieTraceBufSz = aieMemorySz;
-      std::string msg = "Trace buffer size is too big for memory resource.  Using " + std::to_string(aieMemorySz) + " instead." ;
-      xrt_core::message::send(xrt_core::message::severity_level::warning, "XRT", msg);
+
+    // Second, check against amount dedicated as device memory (Linux only)
+#ifndef _WIN32
+    try {
+      std::string line;
+      std::ifstream ifs;
+      ifs.open("/proc/meminfo");
+      while (getline(ifs, line)) {
+        if (line.find("CmaFree") == std::string::npos)
+          continue;
+          
+        std::vector<std::string> cmaVector;
+        boost::split(cmaVector, line, boost::is_any_of(":"));
+        auto deviceMemorySize = std::stoull(cmaVector.at(1)) * 1024;
+        if (deviceMemorySize == 0)
+          break;
+
+        double percentSize = (100.0 * aieTraceBufSize) / deviceMemorySize;
+        if (percentSize >= 80.0) {
+          std::string msg = "Requested AIE trace buffer is " + std::to_string(percentSize) + "% of free device memory."
+              + " You may run into errors (e.g., bad alloc) depending upon the memory usage of your application.";
+          xrt_core::message::send(xrt_core::message::severity_level::warning, "XRT", msg);
```

Comment:
should this be Info?

Reply:
The problem with info is no one will see it. It'll get lost amongst many messages. Since it's getting close to device memory size, I made it a warning.

Reply:
as long as it's just aie it should be fine